plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -123,34 +160,39 @@ public void invalidate(Partition partition)
         if (isCacheEnabledFor(partition.getSchemaTableName()) && isLocationPresent(partition.getStorage())) {
             cache.invalidate(new Path(partition.getStorage().getLocation()));
         }
+        delegate.invalidate(partition);
     }
 
-    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(ValueHolder cachedValueHolder, RemoteIterator<LocatedFileStatus> iterator, Path path)
+    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(ValueHolder cachedValueHolder, Path path)
     {
         return new RemoteIterator<>()
         {
-            private final List<LocatedFileStatus> files = new ArrayList<>();
+            private int fileIndex;
 
             @Override
             public boolean hasNext()
                     throws IOException
             {
-                boolean hasNext = iterator.hasNext();
-                if (!hasNext) {
+                try {
+                    boolean hasNext = cachedValueHolder.getCachedFile(fileIndex).isPresent();
+                    // Update cache weight of cachedValueHolder for a given path.
                     // The cachedValueHolder acts as an invalidation guard. If a cache invalidation happens while this iterator goes over
                     // the files from the specified path, the eventually outdated file listing will not be added anymore to the cache.
-                    cache.asMap().replace(path, cachedValueHolder, new ValueHolder(files));
+                    cache.asMap().replace(path, cachedValueHolder, cachedValueHolder);
+                    return hasNext;
+                }
+                catch (IOException exception) {
+                    // invalidate cached value to force retry of directory listing
+                    cache.invalidate(path);
+                    throw exception;
                 }
-                return hasNext;
             }
 
             @Override
             public LocatedFileStatus next()
                     throws IOException
             {
-                LocatedFileStatus next = iterator.next();
-                files.add(next);
-                return next;
+                return cachedValueHolder.getCachedFile(fileIndex++).orElseThrow();
```

Comment:
`next()` should also update the cache weight. it's not idiomatic, but it's technically possible to use Iterator without ever calling `hasNext()`

it's kind of pre-existing, but not quite. before the change, the `next()` was not increasing cached memory consumption, while after the change it does.